android/runtime/v8/src/native/JSException.cpp

```diff
@@ -12,41 +15,55 @@
 
 #include "JSException.h"
 
+#define MAX_STACK 10
+
 using namespace v8;
 
 namespace titanium {
 
 Local<Value> JSException::fromJavaException(v8::Isolate* isolate, jthrowable javaException)
 {
-	JNIEnv *env = JNIScope::getEnv();
+	JNIEnv* env = JNIScope::getEnv();
 	if (!env) {
 		return GetJNIEnvironmentError(isolate);
 	}
 
-	env->ExceptionDescribe();
-
 	bool deleteRef = false;
 	if (!javaException) {
 		javaException = env->ExceptionOccurred();
-		env->ExceptionClear();
 		deleteRef = true;
 	}
+	env->ExceptionClear();
 
-	//env->ExceptionDescribe();
-
-	jstring message = (jstring) env->CallObjectMethod(javaException, JNIUtil::throwableGetMessageMethod);
-	if (!message) {
+	jstring javaMessage = (jstring) env->CallObjectMethod(javaException, JNIUtil::throwableGetMessageMethod);
+	if (!javaMessage) {
 		return THROW(isolate, "Java Exception occurred");
 	}
+	const char* messagePtr = env->GetStringUTFChars(javaMessage, NULL);
+	std::stringstream message;
+	message << messagePtr;
+	env->ReleaseStringUTFChars(javaMessage, messagePtr);
+	env->DeleteLocalRef(javaMessage);
 
-	Local<Value> jsMessage = TypeConverter::javaStringToJsString(isolate, env, message);
-	env->DeleteLocalRef(message);
+	jobjectArray frames = (jobjectArray) env->CallObjectMethod(javaException, JNIUtil::throwableGetStackTraceMethod);
+	jsize frames_length = env->GetArrayLength(frames);
+	for (int i = 0; i < (frames_length > MAX_STACK ? MAX_STACK : frames_length); i++) {
+		jobject frame = env->GetObjectArrayElement(frames, i);
+		jstring javaStack = (jstring) env->CallObjectMethod(frame, JNIUtil::stackTraceElementToStringMethod);
+
+		const char* stackPtr = env->GetStringUTFChars(javaStack, NULL);
+		message << std::endl << "    " << stackPtr;
+
+		env->ReleaseStringUTFChars(javaStack, stackPtr);
+		env->DeleteLocalRef(javaStack);
+	}
+	message << std::endl;
 
 	if (deleteRef) {
 		env->DeleteLocalRef(javaException);
 	}
 
-	return isolate->ThrowException(jsMessage->ToString(isolate));
+	return isolate->ThrowException(String::NewFromUtf8(isolate, message.str().c_str()));
```

Comment:
So here we have always done the wrong thing which is to throw a String. We should be throwing an actual Exception (Error).

I think the only wrinkle here is to override the `stack` property to hold the java stack trace we generate here.

```cpp
// Grab the top-level error message
Local<Value> jsMessage = TypeConverter::javaStringToJsString(isolate, env, javaMessage);
env->DeleteLocalRef(javaMessage);
// Create a JS Error holding this message
Local<Value> error = v8::Exception::Error(jsMessage);

// Now loop through the java stack and generate a JS String from the result and assign to Local<String> stack
std::stringstream stackStream;
jobjectArray frames = (jobjectArray) env->CallObjectMethod(javaException, JNIUtil::throwableGetStackTraceMethod);
jsize frames_length = env->GetArrayLength(frames);
for (int i = 0; i < (frames_length > MAX_STACK ? MAX_STACK : frames_length); i++) {
	jobject frame = env->GetObjectArrayElement(frames, i);
	jstring javaStack = (jstring) env->CallObjectMethod(frame, JNIUtil::stackTraceElementToStringMethod);

	const char* stackPtr = env->GetStringUTFChars(javaStack, NULL);
 	stackStream << std::endl << "    " << stackPtr;

	env->ReleaseStringUTFChars(javaStack, stackPtr);
	env->DeleteLocalRef(javaStack);
}
stackStream << std::endl;
Local<Value> stack = String::NewFromUtf8(isolate, stackStream.str().c_str());

// Now explicitly assign our properly generated stacktrace
error->Set(context, STRING_NEW(isolate, "stack"), stack);

// throw it
return isolate->ThrowException(error);
```